src/libmcrx.c

```diff
@@ -1055,3 +1076,418 @@ mcrx_is_system_error(enum mcrx_error_code err) {
+  }
+
+  LIST_INSERT_HEAD(&mnatctx->mnats_head, entry, mnat_entries);
+
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnat_ctx_remove_entry:
+ *
+ * Remove the mcrx mnat entry.
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_remove_entry(
+    struct mcrx_mnat_ctx *mnatctx, const char *global_source,
+    const char *global_group) {
+  if (global_source == NULL || global_group == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+
+  struct mcrx_mnat_entry *entry = mcrx_mnat_ctx_find_entry(mnatctx,
+      global_source, global_group);
+  if (entry == NULL) {
+    return MCRX_ERR_NOT_EXISTED;
+  }
+
+  LIST_REMOVE(entry, mnat_entries);
+  free(entry);
+
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnat_ctx_update_entry:
+ *
+ * Update the mcrx mnat entry.
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_update_entry(
+    struct mcrx_mnat_ctx *mnatctx, const char *global_source,
+    const char *global_group, const char *local_source, const char *local_group) {
+  if (global_source == NULL || global_group == NULL || local_source == NULL
+      || local_group == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+
+  struct mcrx_mnat_entry *entry = mcrx_mnat_ctx_find_entry(mnatctx,
+      global_source, global_group);
+  if (entry == NULL) {
+    return MCRX_ERR_NOT_EXISTED;
+  }
+
+  int ret;
+  ret = inet_pton(AF_INET, global_source, &entry->global_addrs.v4.source);
+  if (ret > 0) {
+    ret = inet_pton(AF_INET, global_group, &entry->global_addrs.v4.group);
+    if (ret > 0) {
+      ret = inet_pton(AF_INET, local_source, &entry->local_addrs.v4.source);
+      if (ret > 0) {
+        ret = inet_pton(AF_INET, local_group, &entry->local_addrs.v4.group);
+        if (ret > 0) {
+          entry->addr_type = MCRX_ADDR_TYPE_V4;
+        }
+      }
+    }
+  }
+
+  ret = inet_pton(AF_INET, global_source, &entry->global_addrs.v6.source);
+  if (ret > 0) {
+    ret = inet_pton(AF_INET, global_group, &entry->global_addrs.v6.group);
+    if (ret > 0) {
+      ret = inet_pton(AF_INET, local_source, &entry->local_addrs.v6.source);
+      if (ret > 0) {
+        ret = inet_pton(AF_INET, local_group, &entry->local_addrs.v6.group);
+        if (ret > 0) {
+          entry->addr_type = MCRX_ADDR_TYPE_V6;
+        }
+      }
+    }
+  }
+
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnat_ctx_ref:
+ *
+ **/
+MCRX_EXPORT struct mcrx_mnat_ctx* mcrx_mnat_ctx_ref(
+    struct mcrx_mnat_ctx *mnatctx) {
+  mnatctx->refcount++;
+  return mnatctx;
+}
+
+MCRX_EXPORT struct mcrx_mnat_ctx* mcrx_mnat_ctx_unref(
+    struct mcrx_mnat_ctx *mnatctx) {
+  mnatctx->refcount--;
+  if (mnatctx->refcount > 0) {
+    return mnatctx;
+  }
+
+  while (!LIST_EMPTY(&mnatctx->mnats_head)) {
+    struct mcrx_mnat_entry *entry = LIST_FIRST(&mnatctx->mnats_head);
+    LIST_REMOVE(entry, mnat_entries);
+    free(entry);
+  }
+
+  free(mnatctx);
+  return NULL;
+}
+
+/**
+ * mcrx_mnat_ctx_find_entry_from_subscription:
+ *
+ * Retrieve the mcrx mnat entry based on subscription.
+ *
+ **/
```

Comment:
this function exposes the mnat_entry object to callers, requiring the ref count, do you think we need it, or do we just need the currently mapped addresses?